runtimevar/blobvar/blobvar.go

```diff
@@ -34,6 +51,90 @@ import (
 	"gocloud.dev/runtimevar/driver"
 )
 
+func init() {
+	runtimevar.DefaultURLMux().RegisterVariable(Scheme, &lazyBucketOpener{})
+}
+
+// Scheme is the URL scheme blobvar registers its URLOpener under on runtimevar.DefaultMux.
+const Scheme = "blob"
+
+type lazyBucketOpener struct {
+	mu      sync.Mutex
+	buckets map[string]*blob.Bucket
+}
+
+func (o *lazyBucketOpener) OpenVariableURL(ctx context.Context, u *url.URL) (*runtimevar.Variable, error) {
+	bucketURL := u.Query().Get("bucket")
+	if bucketURL == "" {
+		return nil, fmt.Errorf("open variable %q: URL parameter \"bucket\" is required", u)
+	}
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	if o.buckets == nil {
+		o.buckets = map[string]*blob.Bucket{}
+	}
+	bucket := o.buckets[bucketURL]
+	if bucket == nil {
+		var err error
+		bucket, err = blob.OpenBucket(ctx, bucketURL)
+		if err != nil {
+			return nil, fmt.Errorf("open variable %q: OpenBucket failed: %v", u, err)
+		}
+		o.buckets[bucketURL] = bucket
+	}
+	opener := URLOpener{Bucket: bucket}
+	return opener.OpenVariableURL(ctx, u)
+}
+
+// URLOpener opens Variable URLs like "blob://myblobkey?decoder=string".
+// It supports the URL parameters:
+//   - decoder: The decoder to use. Defaults to URLOpener.Decoder, or
+//       runtimevar.BytesDecoder if URLOpener.Decoder is nil.
+//       See runtimevar.DecoderByName for supported values.
+//   - wait: The poll interval; supported values are from time.ParseDuration.
+//       Defaults to 30s.
+type URLOpener struct {
+	// The Bucket to use; required.
+	Bucket *blob.Bucket
+
+	// Decoder and Options can be specified at URLOpener construction time,
+	// or provided/overridden via URL parameters.
+	Decoder *runtimevar.Decoder
+	Options Options
+}
+
+// OpenVariableURL opens the variable at the URL's path. See the package doc
+// for more details.
+func (o *URLOpener) OpenVariableURL(ctx context.Context, u *url.URL) (*runtimevar.Variable, error) {
+	if o.Bucket == nil {
+		return nil, fmt.Errorf("open variable %q: a bucket is required", u)
+	}
+	q := u.Query()
+	if decoderName := q.Get("decoder"); decoderName != "" || o.Decoder == nil {
+		var err error
+		o.Decoder, err = runtimevar.DecoderByName(decoderName)
+		if err != nil {
+			return nil, fmt.Errorf("open variable %q: invalid \"decoder\": %v", u, err)
+		}
+	}
+	for param, values := range q {
+		val := values[0]
+		switch param {
+		case "decoder", "bucket":
```

Comment:
"decoder" can be processed here instead of above.

Reply:
It can't, at least not easily, because of the `|| o.Decoder == nil` clause above; if there is no `Decoder` provided in the `URLOpener`, we want to get the default decoder (by passing "" to `runtimevar.DecoderByName`). In this loop, if there's no decoder param, we'll never call it.